File: 0525-Contiguous-Array.js
var findMaxLength = function(nums) {
    let maxLen = 0;
    let currentSum = 0;
    let sumLocation = new Map();
    sumLocation.set(0, -1);

    for (let i = 0; i < nums.length; i++) {
        currentSum += 2 * nums[i] - 1;
        if(sumLocation.has(currentSum)) {
            maxLen = Math.max(maxLen, i - sumLocation.get(currentSum));
        } else {
            sumLocation.set(currentSum, i)
        }
    }

    return maxLen;
};

File: 0525-Contiguous-Array.py
class Solution(object):
    def findMaxLength(self, nums):
        currentSum = 0
        sumLocation = defaultdict(int)
        sumLocation[0] = -1
        maxLen = 0

        for i, num in enumerate(nums):
            currentSum += (1 if num == 1 else -1)
            if currentSum in sumLocation:
                maxLen = max(maxLen, i - sumLocation[currentSum])
            else:
                sumLocation[currentSum] = i

        return maxLen
